README.md
Ubuntu Lamp Template
===================

Intro
----------
  This cookbook installs and configures everything required to run Drupal, Wordpress, Joomla, or any other PHP application.

  Core Stack:

  - Ubuntu 14.04 LTS - This is the _only_ supported platform
  - PHP5.6 via ppa:ondrej/php5-5.6
  - Apache 2.4 via ppa:ondrej/apache2
  - A very simple Nginx recipe if you'd like LNMP instead
  - MariaDB Galera 10 via http://ftp.osuosl.org/pub/mariadb/repo/10.0/ubuntu in a Galera Cluster
  - Varnish 4.0 from http://repo.varnish-cache.org
  - HAProxy 1.5 from ppa:vbernat/haproxy-1.5
  - Redis 2.8 via the redisio cookbook

  Sites and databases are configured in ```recipes/site_NAME.rb``` and ```recipes/databases.rb``` respectively. ```recipes/filesystems.rb``` sets up gluster mounts.

  Most common tweaks and changes will take place directly in the template. For instance, ```templates/default/apache2/apache2.conf.erb``` and ```templates/default/php/php.ini.erb``` are exactly what you'd expect.

  Some templates are heavily customized. See ```templates/default/mysql/my.cnf.erb``` - Many settings are dynamically set based on systems available memory. Please read and understand ```recipes/_mysql.rb``` before making big changes to that template.

  It's also important to note that this cookbook REQUIRES a cloud network. It will not only assume to look for an 'eth2' (check default.rb) on initial bootstrap, but it opens the firewall COMPLETELY for that interface. See the bootstrap example for how to bootstrap a server with a cloud network attached. The cloud network is assumed to be named the same as the chef environment. You can use servicenet if you want, but this cookbook will not deal with the complexity of firewalling on a shared network. It will whitelist other nodes in the environment, but bootstrapping Galera and Gluster will not work unless you manually whitelist a new servers IP before bootstrapping it. Using a Cloud Network (or any private network) is by far preferred.

Philosopy
----------
  Unlike others, who settle for PHP applications being riddled with single points of failure, this cookbook aims to provide highly available solutions to _all_ common PHP problems, include: memory/session store (HA redis), disk (gluster), database (mysql master-master Galera replication).

  While this mission is far from complete (and obviously ambitious), we're most of the way there.

Bootstrapping new nodes
----------
### Using knife rackspace:
    knife rackspace server create -N"dev-1" -r"role[app]" -E"dev" -I"598a4282-f14b-4e50-af4c-b3e52749d9f9" -f"performance1-4" --no-tcp-test-ssh --ssh-wait-timeout 60 -S"dev-1" --secret-file path/to/secret --bootstrap-version 11.16.4 --network "dev" --rackspace-version v2

### Bootstrap
    knife bootstrap IP_ADDR -N"dev-2" -r"role[app]" -E"dev" --secret-file path/to/secret --bootstrap-version 11.16.4 --network "dev" --rackspace-version v2

Bootstrapping a new environment
----------
Take a peek at "dev" and make note of the set MySQL password and chef version. Use this on any new environment.

### The database
You'll also want to run mysql_secure_installation on MySQL masters as well. Automation is no replacement for a properly setup MySQL master! Assuming you're making use of Galera, DB syncing happens automagically!

### Redis
Each application node (role_app) will have its own redis instance, but a single small shared redis instance must exist for PHP sessions to be shared properly.

Mark one server with a "redis_master" role or checkout the default recipe to override the address_map attribute.

### The static assets
You can use the cloudsink utility: https://github.com/erulabs/cloudsink

An example of syncing all the static files in Drupal (should be carefully studied) from an existing file server to Cloud Files.

Note that performance is _far_ greater when uploading to a NON CDN container than to a CDN container. You can enable CDN after the fact - so if doing an initial build, ALWAYS use a private container for the initial upload (and enable CDN after the fact).

```
    cloudsink -S -t CONTAINER -u USERNAME -k API_KEY -r IAD -F -s sites -f "*.js"
    cloudsink -S -t CONTAINER -u USERNAME -k API_KEY -r IAD -F -s sites -f "*.css"
    cloudsink -S -t CONTAINER -u USERNAME -k API_KEY -r IAD -F -s modules -f "*.js"
    cloudsink -S -t CONTAINER -u USERNAME -k API_KEY -r IAD -F -s modules -f "*.css"
    cloudsink -S -t CONTAINER -u USERNAME -k API_KEY -r IAD -F -s sites/default/files
```

Wordpress
----------
Am example recipe at "site_wordpress.rb" has been provided which deploys a fake wordpress application. The wp-config template is included, with some hints, although you'll almost certainly not want to use this example one (most likely a real wordpress app will have it's own wp-config).

As with all other PHP applications, the real key to performance is to use Redis above all other resources. Every query shipped to Redis instead of to MySQL is a huge win. Even compaired to Memcached.

I strongly recommend using the Redis object cache available here: https://github.com/alleyinteractive/wp-redis

You can use a redis full page cache (as opposed to heavily using Varnish) if you'd like, but I have not vetted this plugin as thourghly as the object cache plugin: https://github.com/BenjaminAdams/wp-redis-cache

Other things to note about wordpress is that you should closely monitor the number of plugins which hammer the database - in particular applications which log to the database. Disable as many as humanly possible.

After this, install W3 Total Cache so that you can use a CDN for your assets. CloudSink (see above) can be used to ship initial loads of files to Cloud Files.

Drupal
----------
Again, simply Redis as many things as humanly possible. The core redis module works fairly well: https://www.drupal.org/project/redis

For CDN, The Rackspace Cloud Files plugin is good, but needs a ton of patches. I've done the lions share of this work here: https://github.com/erulabs/drupal7_cloud_files - but you should check the "issues" list here from time to time: https://www.drupal.org/project/cloud_files


CHANGELOG.md
ubuntu_lamp_template CHANGELOG
==============================

This file is used to list changes made in each version of the ubuntu_lamp_template cookbook.

1.0.0
-----
- James Cobb - Need to bump to 1.0 for production as well as add logrotate.


0.1.0
-----
- [your_name] - Initial release of ubuntu_lamp_template

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/_mysql.rb
# The MySQL galera server recipe, which runs MariaDB 10.0

%w(debconf-utils percona-xtrabackup rsync).each do |pkg|
  package pkg do
    action :install
  end
end

# Tune MySQL automagically
open_files_limit = 8192
mysql_listen_address = node['address_map']['my_ip']

# Default to assuming our server is a 1gb slice
system_total_mem_mb = '1024'
unless node['memory']['total'].nil?
  system_total_mem_mb = (node['memory']['total'].gsub(/kB/, '').to_i / 1024).round
end

# The service is now totally installed and configured - time to configure the cluster
node.default['mariadb']['debian']['user'] = 'root'
node.default['mariadb']['debian']['password'] = node['mysql']['server_root_password']
node.default['mariadb']['debian']['host'] = mysql_listen_address

begin
  resources(directory: '/var/cache/local/preseeding')
rescue Chef::Exceptions::ResourceNotFound
  directory '/var/cache/local/preseeding' do
    owner 'root'
    group 'root'
    mode '0755'
    recursive true
  end
end
template '/var/cache/local/preseeding/mariadb-galera-server.seed' do
  source 'mysql/server-seed.erb'
  owner 'root'
  group 'root'
  mode '0600'
  variables(package_name: 'mariadb-galera-server')
  notifies :run, 'execute[preseed mariadb-galera-server]', :immediately
end
execute 'preseed mariadb-galera-server' do
  command '/usr/bin/debconf-set-selections /var/cache/local/preseeding/mariadb-galera-server.seed'
  action :nothing
end

# Add the MariaDB 10 repo
apt_repository 'mariadb' do
  uri 'http://ftp.osuosl.org/pub/mariadb/repo/10.0/ubuntu'
  distribution node['lsb']['codename']
  components ['main']
  # Force use of port 80 for keyserver transaction
  # this bypasses issues with strict firewalls on testing machines
  keyserver 'hkp://keyserver.ubuntu.com:80'
  key '0xcbcb082a1bb943db'
end

# Install mariadb server!
package 'mariadb-galera-server' do
  action [:upgrade, :install]
end

# MySQL ulimit setting
user_ulimit 'mysql' do
  filehandle_limit open_files_limit
  core_hard_limit 'unlimited'
end

template '/root/.my.cnf' do
  source 'mysql/.my.cnf.erb'
  owner 'root'
  group 'root'
  mode '0600'
  variables(
    pass: node['mysql']['server_root_password']
  )
end

galera_options = {}
gcomm = ''
unless Chef::Config[:solo]
  first = true
  db_galera_raw = search(:node, "tags:*galera_node* AND chef_environment:#{node.chef_environment}")
  if !db_galera_raw.nil? && !db_galera_raw.empty? && !db_galera_raw.first.nil?
    db_galera_raw.each do |raw_node|
      next if raw_node.name == node.name
      if node['address_map']['use_service_net']
        best_ip = raw_node['network']['interfaces'][node['address_map']['service_net_iface']]['addresses'].find {|addr, addr_info| addr_info[:family] == 'inet'}.first
      else
        privatenet = raw_node['rackspace']['private_networks'].find { |netwrk| netwrk['label'] == node.chef_environment }
        best_ip = privatenet['ips'].first['ip']
      end
      next if best_ip.nil?
      gcomm += ',' unless first
      gcomm += best_ip
      first = false
    end
  end
end

galera_options['wsrep_cluster_address'] = gcomm
galera_options['wsrep_cluster_name'] = "galera_cluster_#{node.chef_environment}"
galera_options['wsrep_sst_method'] = 'rsync'

template '/etc/mysql/my.cnf' do
  source 'mysql/my.cnf.erb'
  owner 'root'
  group 'root'
  mode '0600'
  variables(
    listen: mysql_listen_address,
    server_id: mysql_listen_address.gsub(/\./, ''),
    open_files_limit: 65_535,
    table_open_cache: (open_files_limit / 4).round,
    # innodb_buffer_pool_size will be 50% of the systems total memory
    innodb_buffer_pool_size: (system_total_mem_mb / 3).round,
    innodb_log_file_size: (system_total_mem_mb / 32).round,
    innodb_log_buffer_size: (system_total_mem_mb / 32).round,
    # MariaDB does not need a very large connection pool - particularly when paired with PHP 5.6
    # which is quite good at closing connections - we'll allow 15 connections per GB of memory for now
    # which is fairly arbitrary, but makes an OK baseline
    max_connections: ((system_total_mem_mb / 1024).round + 1) * 15,
    tmp_table_size: (system_total_mem_mb / 32).round,
    query_cache_size: (system_total_mem_mb / 32).round,
    galera_options: galera_options
  )
  notifies :restart, 'service[mysql]'
end

# Ensure service is runing
service 'mysql' do
  service_name 'mysql'
  supports restart: true, status: true
  action [:enable, :start]
  retries 1
  retry_delay 3
  # notifies :create, 'template[/root/constants.sql]', :immediate
end

execute 'correct-debian-grants' do
  command 'mysql -r -B -N -e "GRANT SELECT, INSERT, UPDATE, DELETE, ' + \
    'CREATE, DROP, RELOAD, SHUTDOWN, PROCESS, FILE, REFERENCES, INDEX, ' + \
    'ALTER, SHOW DATABASES, SUPER, CREATE TEMPORARY TABLES, ' + \
    'LOCK TABLES, EXECUTE, REPLICATION SLAVE, REPLICATION CLIENT, ' + \
    'CREATE VIEW, SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, ' + \
    "CREATE USER, EVENT, TRIGGER ON *.* TO '" + \
    node['mariadb']['debian']['user'] + \
    "'@'" + node['mariadb']['debian']['host'] + "' IDENTIFIED BY '" + \
    node['mariadb']['debian']['password'] + "' WITH GRANT OPTION\""
  action :run
  only_if do
    cmd = Mixlib::ShellOut.new("/usr/bin/mysql --user=\"" + \
      node['mariadb']['debian']['user'] + \
      "\" --password=\"" + node['mariadb']['debian']['password'] + \
      "\" -r -B -N -e \"SELECT 1\"")
    cmd.run_command
    cmd.error?
  end
  ignore_failure true
end

tag('mysql_master')
tag('galera_node')
